File: zadanie.m
global n h;
n = input('Podaj n: ');
h = 2/n;

function result = e (i,x)
global n h;
if x >= h*(i - 2) & x < h*(i -1)
    result = (x - h*(i - 2))/h;
elseif x >= h*(i - 1) & x < h*(i)
    result = (h - (x - h*(i -1 )))/h;
else
    result = 0;

end
end

function result = e_prim(i,x)
global n h;
if x >= h*(i - 2) & x < h*(i -1)
    result = 1/h;
elseif x >= h*(i - 1) & x < h*(i)
    result = -1/h;
else
    result = 0;

end
end

function result = L(i)
global n h;
result = -37/2 * e(i,0) + 3/2 * e(i,1);
end

function result = k(x)
global n h;
if x <= 1 & x >= 0
    result = 1;
elseif x <= 2 & x > 1
    result = 2;
else
    result = 0;
end
end

function result = B(i,j)
global n h;
result = -2*e(j,2)*e_prim(i,2)-e(j,0)*e(i,0)+integral(@(x) k(x).*e_prim(i,x).*e_prim(j,x),0,2,'ArrayValued', true);
end

function result = u(W,x)
global n h;
result = 3/2*x;
for i = 1:n
    result = result + W(i)*e(i,x);
end
end

A = zeros(n, n);
for i = 1:n
    for j = 1:n
        A(i, j) = B(i, j);
    end
end

V = zeros(n, 1);
for i = 1:n
    V(i) = L(i);
end

W = A\V;

x_values = linspace(0, 2, 1000);
u_values = arrayfun(@(x) u(W, x), x_values);

figure;
plot(x_values, u_values);
xlabel('x');
ylabel('u(x)');
title('Plot of the function u(x)');
grid on;

File: zadanie.py
import numpy as np
from scipy.integrate import quad
import matplotlib.pyplot as plt
import seaborn as sns


n = int(input("daj n "))
h = 2 / n


def e(i, x):
    if x >= h * (i - 1) and x <= h * i:
        return (x - h * (i - 1)) / h * i
    elif x >= h * i and x <= h * (i + 1):
        return (h * (i + 1) - x) / h * i
    else:
        return 0


def e_prim(i, x):
    if x >= h * (i - 1) and x <= h * i:
        return 1 / h
    elif x >= h * i and x <= h * (i + 1):
        return -1 / h
    else:
        return 0


def L(v):
    return quad(lambda x: 100 * (x**2) * e(v, x), 0, 2)[0] - 40 * e(v, 0)


def k(x):
    if x >= 0 and x <= 1:
        return 1
    elif x > 1 and x <= 2:
        return 2 * x


def B(u, v):
    return quad(lambda x: k(x) * e_prim(u, x) * e_prim(v, x), 0, 2)[0] - e(u, 0) * e(
        v, 0
    )


def u(W, x):
    sum_val = -20
    for i in range(n):
        sum_val += W[i] * e(i, x)
    return sum_val


A = [[0 for _ in range(n)] for _ in range(n)]

for i in range(len(A)):
    for j in range(len(A[i])):
        A[i][j] = B(i, j)

V = [0 for _ in range(n)]

for i in range(len(V)):
    V[i] = L(i)

W = np.linalg.solve(A, V)

x_vals = np.linspace(0, 2, 1000)
y_vals = []
for x in x_vals:
    y_vals.append(u(W, x))

print(y_vals)

sns.set_style(style="darkgrid")
sns.lineplot(x=x_vals, y=y_vals, label="u(W, x)")
plt.xlim(0, 2)
plt.xlabel("x")
plt.ylabel("u(W, x)")
plt.title("Plot of u(W, x)")
plt.legend()
plt.show()
